<a href="https://colab.research.google.com/github/muhammaduns/LearnAI/blob/main/MNIST_Practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

** Loading the MNIST dataset in Keras**

In [1]:
from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11490434/11490434 [==============================] - 2s 0us/step


In [2]:
train_images.shape

(60000, 28, 28)

In [3]:
test_images.shape

(10000, 28, 28)

In [4]:
len(train_labels), len(train_images)

(60000, 60000)

In [5]:
 len(test_labels), len(test_images)

(10000, 10000)

In [6]:
print(len(train_images))

60000


In [7]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

In [8]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

The network architecture

imporintg layers

taking model either sequential 
and define layers with activation

---------------------
Compilation Step
---------------------

 To make the model ready for training, we need to pick three more things as part of
the compilation step:

 An optimizer—The mechanism through which the model will update itself based
on the training data it sees, so as to improve its performance.

 A loss function—How the model will be able to measure its performance on the
training data, and thus how it will be able to steer itself in the right direction.

 Metrics to monitor during training and testing—Here, we’ll only care about accuracy (the fraction of the images that were correctly classified).


In [9]:
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([
 layers.Dense(512, activation="relu"),
 layers.Dense(10, activation="softmax")
])

The compilation step

In [10]:
model.compile(optimizer="rmsprop",
 loss="sparse_categorical_crossentropy",
 metrics=["accuracy"])

Before training, we’ll preprocess the data by reshaping it into the shape the model
expects and scaling it so that all values are in the [0, 1] interval. Previously, our training images were stored in an array of shape (60000, 28, 28) of type uint8 with values
in the [0, 255] interval. We’ll transform it into a float32 array of shape (60000, 28 *
28) with values between 0 and 1.

 Preparing the image data

In [11]:
train_images = train_images.reshape((60000, 28 * 28))
train_images.shape

(60000, 784)

In [12]:
train_images = train_images.astype("float32") / 255
print(train_images.shape)
print(len(train_images))

(60000, 784)
60000


In [13]:
test_images = test_images.reshape((10000, 28 * 28))
print(test_images.shape)
print(len(test_images))

(10000, 784)
10000


In [14]:
test_images = test_images.astype("float32") / 255

 “Fitting” the model

```
# This is formatted as code
```
We’re now ready to train the model, which in Keras is done via a call to the model’s
fit() method—we fit the model to its training data.


In [15]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 5s 3ms/step - loss: 0.2570 - accuracy: 0.9249
Epoch 2/5
469/469 [==============================] - 1s 3ms/step - loss: 0.1038 - accuracy: 0.9694
Epoch 3/5
469/469 [==============================] - 1s 3ms/step - loss: 0.0693 - accuracy: 0.9800
Epoch 4/5
469/469 [==============================] - 1s 3ms/step - loss: 0.0495 - accuracy: 0.9850
Epoch 5/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0376 - accuracy: 0.9887


Using the model to make predictions

In [16]:
test_digits = test_images[0:10]

In [17]:
predictions = model.predict(test_digits)

1/1 [==============================] - 0s 102ms/step


In [18]:
predictions[0]

array([4.9831104e-09, 4.0576723e-10, 1.1063878e-06, 2.9102303e-04,
       3.8216885e-13, 6.2330578e-08, 4.1091982e-15, 9.9970704e-01,
       5.0498215e-08, 6.8932587e-07], dtype=float32)

In [19]:
predictions[0].argmax()

7

In [20]:
predictions[2], predictions[2].argmax()

(array([2.8725232e-08, 9.9934214e-01, 4.8797214e-05, 4.5800512e-06,
        1.2924876e-04, 3.8618027e-06, 4.4999542e-06, 2.9128353e-04,
        1.7442470e-04, 1.1797596e-06], dtype=float32), 1)

In [21]:
predictions[3], predictions[3].argmax()

(array([9.9996698e-01, 4.4978035e-11, 2.8099082e-06, 4.9587491e-08,
        5.3760431e-08, 3.3519777e-07, 9.6553622e-06, 1.5292635e-05,
        3.1316547e-10, 4.8226448e-06], dtype=float32), 0)

In [22]:
for i in range(10):
  print(predictions[i], predictions[i].argmax())

[4.9831104e-09 4.0576723e-10 1.1063878e-06 2.9102303e-04 3.8216885e-13
 6.2330578e-08 4.1091982e-15 9.9970704e-01 5.0498215e-08 6.8932587e-07] 7
[1.3570008e-10 6.0564986e-07 9.9999940e-01 6.4412981e-10 2.9341482e-19
 1.3213888e-09 3.9070686e-10 1.1531577e-17 2.4261309e-09 2.4596029e-17] 2
[2.8725232e-08 9.9934214e-01 4.8797214e-05 4.5800512e-06 1.2924876e-04
 3.8618027e-06 4.4999542e-06 2.9128353e-04 1.7442470e-04 1.1797596e-06] 1
[9.9996698e-01 4.4978035e-11 2.8099082e-06 4.9587491e-08 5.3760431e-08
 3.3519777e-07 9.6553622e-06 1.5292635e-05 3.1316547e-10 4.8226448e-06] 0
[9.8625651e-07 4.1306642e-11 3.2254820e-06 9.8376791e-09 9.9301130e-01
 1.6836619e-08 1.0098008e-06 1.8331113e-05 3.0654454e-07 6.9648097e-03] 4
[2.0702946e-09 9.9907148e-01 5.4702946e-07 7.3547005e-07 8.1000144e-05
 2.4538567e-08 2.9381086e-08 8.3416846e-04 1.1536989e-05 4.2397568e-07] 1
[4.50507097e-11 3.80855489e-08 8.69397425e-08 2.30080648e-08
 9.99556005e-01 6.46606736e-07 1.01702774e-07 4.76271707e-05
 3.38238

In [24]:
#  Evaluating the model on new data
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 [==============================] - 1s 2ms/step - loss: 0.0675 - accuracy: 0.9792


In [25]:
test_loss, test_acc

(0.0674821138381958, 0.979200005531311)

The test-set accuracy turns out to be 97.8%—that’s quite a bit lower than the trainingset accuracy (98.9%). This gap between training accuracy and test accuracy is an
example of overfitting: the fact that machine learning models tend to perform worse
on new data than on their training data. Overfitting is a central topic in chapter 3.